WebDriverAgentLib/Categories/XCUIApplication+FBLaunch.m

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import "XCUIApplication+FBLaunch.h"
+
+#import <objc/runtime.h>
+
+#import "FBAlert.h"
+#import "FBLogger.h"
+#import "XCUIapplication.h"
+#import "XCUIApplication+FBAlert.h"
+#import "XCUIApplication+FBHelpers.h"
+
+@implementation XCUIApplication (FBLaunch)
+
+static char XCUIAPPLICATION_DID_START_WO_BLOCKING_ALERT;
+
+@dynamic fb_didStartWithoutBlockingAlert;
+
+- (void)setFb_didStartWithoutBlockingAlert:(NSNumber *)didStartWithoutBlockingAlert
+{
+  objc_setAssociatedObject(self, &XCUIAPPLICATION_DID_START_WO_BLOCKING_ALERT, 
+                           didStartWithoutBlockingAlert, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+}
+
+- (NSNumber *)fb_didStartWithoutBlockingAlert
+{
+  return objc_getAssociatedObject(self, &XCUIAPPLICATION_DID_START_WO_BLOCKING_ALERT);
+}
+
+- (void)fb_scheduleNextDispatchWithInterval:(NSTimeInterval)interval
+                                timeStarted:(uint64_t)timeStarted
+                                    timeout:(NSTimeInterval)timeout
+                              exceptionName:(NSString *)exceptionName
+{
+  dispatch_time_t dispatchTime = dispatch_time(DISPATCH_TIME_NOW, (int64_t) (interval * NSEC_PER_SEC));
+  dispatch_after(dispatchTime, dispatch_get_main_queue(), ^{
+    NSTimeInterval duration = (clock_gettime_nsec_np(CLOCK_MONOTONIC_RAW) - timeStarted) / NSEC_PER_SEC;
+    if ([self.fb_didStartWithoutBlockingAlert boolValue] || duration > timeout) {
+      return;
+    }
+
+    FBAlert *alert = nil;
+    @try {
+      XCUIElement *alertElement = [self.class.fb_systemApplication fb_alertElement];
+      if (nil != alertElement) {
+        alert = [FBAlert alertWithElement:alertElement];
+      }
+    } @catch (NSException *e) {
+      [FBLogger logFmt:@"Got an unexpected exception while checking for system alerts: %@\n%@", e.reason, e.callStackSymbols];
+    }
+    if (nil != alert) {
+      NSString *reason = [NSString stringWithFormat:@"The application '%@' cannot be launched because it is blocked by an unexpected system alert: %@", self.bundleID, alert.text];
+      @throw [NSException exceptionWithName:exceptionName reason:reason userInfo:nil];
```

Comment:
Do we need to stop WDA process here? Maybe here can be just an error log?

I tested an invalid app that fails to launch by `[self launch];` and the view has an error message. So the error response is by this `[self launch];` as invalid app. Then, this alert log is also shown in the system log, but then the WDA process ends. Technically we can keep running WDA process in this case to reduce process relaunch for the next session

Reply:
The flow is:

WDA create session handler -> app launch
And during the second phase is running we also run our async alert checker. 
The question is how we could then interrupt the `[launch]` call without throwing an excepting in the async checker and interrupting WDA itself? Btw, what happens if you put `[self terminate]` call there instead of throwing the exception?

Reply:
changed it to use a "safer" flow, but not sure if this would actually work

Reply:
Still I can't find an alert that blocks launch call. Invalid signature alert didn't block the launch api (the launch api raise an exception with invalid signature error message and the system alert remains), so I wondered if simply we can check system alert existence in try/catch of the launch app call.

Let me check the terminate, and keep finding blocking alert behavior tomorrow

Reply:
>Invalid signature alert didn't block the launch api

The signature should be valid, just an enterprise one. You can install such apps to any device, but it is necessary to confirm the profile in settings before an app could be executed.

![scr](https://support.apple.com/library/content/dam/edam/applecare/images/en_US/iOS/iphone6-ios9-enterprise-untrusted_enterprise_app.png)

Reply:
`[self terminate];` looks good than crash.
I don't have good enterprise developer account, but `Untrusted Developer` which also needs to trust it via settings app did not block in `launch`. The `[self launch]` immediately raised an exception with `Error Domain=com.facebook.WebDriverAgent Code=1 "The application 'com.kazucocoa.testUITests' cannot be launched because of an unexpected error: Error Domain=FBSOpenApplicationServiceErrorDomain Code=1 `... message.

> Selenium::WebDriver::Error::UnknownError: An unknown server-side error occurred while processing the command. Original error: Unable to start WebDriverAgent session because of an unexpected failure: A new session could not be created. Details: Error Domain=com.facebook.WebDriverAgent Code=1 "The application 'com.kazucocoa.testUITests' cannot be launched because of an unexpected error: Error Domain=FBSOpenApplicationServiceErrorDomain Code=1 "The request to open "com.kazucocoa.testUITests" failed." UserInfo={BSErrorCodeDescription=RequestDenied, NSUnderlyingError=0x302577a50 {Error Domain=FBSOpenApplicationErrorDomain Code=3 "Unable to launch com.kazucocoa.testUITests because it has an invalid code signature, inadequate entitlements or its profile has not been explicitly trusted by the user." UserInfo={BSErrorCodeDescription=Security, NSLocalizedFailureReason=Unable to launch com.kazucocoa.testUITests because it has an invalid code signature, inadequate entitlements or its profile has not been explicitly trusted by the user.}}, NSLocalizedDescription=The request to open "com.kazucocoa.testUITests" failed., FBSOpenApplicationRequestID=0x753d, NSLocalizedFailureReason=The request was denied by service delegate (SBMainWorkspace) for reason: Security ("Unable to launch com.kazucocoa.testUITests because it has an invalid code signature, inadequate entitlements or its profile has not been explicitly trusted by the user").}" UserInfo={NSLocalizedDescription=The application 'com.kazucocoa.testUITests' cannot be launched because of an unexpected error: Error Domain=FBSOpenApplicationServiceErrorDomain Code=1 "The request to open "com.kazucocoa.testUITests" failed." UserInfo={BSErrorCodeDescription=RequestDenied, NSUnderlyingError=0x302577a50 {Error Domain=FBSOpenApplicationErrorDomain Code=3 "Unable to launch com.kazucocoa.testUITests because it has an invalid code signature, inadequate entitlements or its profile has not been explicitly trusted by the user." UserInfo={BSErrorCodeDescription=Security, NSLocalizedFailureReason=Unable to launch com.kazucocoa.testUITests because it has an invalid code signature, inadequate entitlements or its profile has not been explicitly trusted by the user.}}, NSLocalizedDescription=The request to open "com.kazucocoa.testUITests" failed., FBSOpenApplicationRequestID=0x753d, NSLocalizedFailureReason=The request was denied by service delegate (SBMainWorkspace) for reason: Security ("Unable to launch com.kazucocoa.testUITests because it has an invalid code signature, inadequate entitlements or its profile has not been explicitly trusted by the user").}}

Reply:
Thanks a lot for checking it.